nemo_aligner/models/mm/stable_diffusion/megatron_sd_draftp_model.py

```diff
@@ -0,0 +1,421 @@
+                .contiguous()
+                .view(-1, *intermediates["pred_x0"][0].shape[1:])
+            )
+
+            vae_decoder_output = []
+            idx_denoised_imgs = [self.inference_steps + (self.inference_steps + 1) * i for i in range(batch_size)]
+
+            for i in range(0, batch_size, self.vae_batch_size):
+                image = self.model.differentiable_decode_first_stage(
+                    trajectories_predx0[idx_denoised_imgs[i : i + self.vae_batch_size]]
+                )
+
+                vae_decoder_output.append(image)
+
+            vae_decoder_output = torch.cat(vae_decoder_output, dim=0)
+            vae_decoder_output = torch.clip((vae_decoder_output + 1) / 2, 0, 1) * 255.0
+
+            log_reward = [
+                self.reward_model.get_reward(
+                    vae_decoder_output[i].unsqueeze(0).detach().permute(0, 2, 3, 1), batch
+                ).item()
+                for i in range(batch_size)
+            ]
+            log_img = [
+                np.transpose(vae_decoder_output[i].float().detach().cpu().numpy(), (1, 2, 0))
+                for i in range(batch_size)
+            ]
+            return log_img, log_reward
+
+    @torch.no_grad()
+    def log_visualization(self, prompts):
+
+        batch_size = len(prompts)
+        latents = torch.randn(
+            [
+                batch_size,
+                self.in_channels,
+                self.height // self.downsampling_factor,
+                self.width // self.downsampling_factor,
+            ],
+            generator=None,
+        ).to(torch.cuda.current_device())
+
+        image_draft_p, reward_draft_p = self.generate_log_images(latents, prompts, self.model)
+        image_init, reward_init = self.generate_log_images(latents, prompts, self.init_model)
+
+        images = []
+        captions = []
+        for i in range(len(image_draft_p)):
+            images.append(image_draft_p[i])
+            images.append(image_init[i])
+            captions.append("DRaFT+: " + prompts[i] + ", Reward = " + str(reward_draft_p[i]))
+            captions.append("SD: " + prompts[i] + ", Reward = " + str(reward_init[i]))
+
+        self.wandb_logger.loggers[1].log_image(
+            key="Inference Images",
+            images=[wandb.Image(Image.fromarray(img.round().astype("uint8"))) for img in images],
+            caption=captions,
+        )
+
+    def generate(
+        self, batch, x_T,
+    ):
+
+        # get autocast_dtype
+        if self.cfg.precision == "bf16":
+            autocast_dtype = torch.bfloat16
+        elif int(self.cfg.precision) == 32:
+            autocast_dtype = torch.float
+        elif int(self.cfg.precision) == 16:
+            autocast_dtype = torch.half
+        else:
+            raise ValueError('precision must be in [32, 16, "bf16"]')
+
+        with torch.cuda.amp.autocast(
+            enabled=autocast_dtype in (torch.half, torch.bfloat16), dtype=autocast_dtype,
```

Comment:
does this model not support amp_O2=True? in amp_O2 = True we don't usually need any autocast, but here is still autocasts

Reply:
it doesn't, check [here](https://github.com/NVIDIA/NeMo/blob/0fbfa211ee6681ff1937e0fe902d5460a0c5a342/nemo/collections/multimodal/models/text_to_image/stable_diffusion/ldm/ddpm.py#L1676), also it is not initialize in the config so it is set to false